py/src/braintrust/oai.py

```diff
@@ -1,41 +1,186 @@
+                                    }
+                                )
+                                first = False
+                            all_results.append(item if isinstance(item, dict) else item.dict())
+                            yield item
+                        span.log(output=all_results)
+                    finally:
+                        span.end()
+
+                should_end = False
+                return gen()
+            else:
+                log_response = raw_response if isinstance(raw_response, dict) else raw_response.dict()
+                span.log(
+                    metrics={
+                        "tokens": log_response["usage"]["total_tokens"],
+                        "prompt_tokens": log_response["usage"]["prompt_tokens"],
+                        "completion_tokens": log_response["usage"]["completion_tokens"],
+                    },
+                    output=log_response["choices"][0],
+                )
+                return raw_response
+        finally:
+            if should_end:
+                span.end()
+
+    @classmethod
+    def _parse_params(cls, params):
+        params = {**params}
+        messages = params.pop("messages", None)
+        return {
+            "input": messages,
+            "metadata": params,
+        }
+
+
+class ChatCompletionV0Wrapper(NamedWrapper):
+    def __init__(self, chat):
+        self.__chat = chat
+        super().__init__(chat)
+
+    def create(self, *args, **kwargs):
+        return ChatCompletionWrapper(self.__chat.create, self.__chat.acreate).create(*args, **kwargs)
+
+    async def acreate(self, *args, **kwargs):
+        return await ChatCompletionWrapper(self.__chat.create, self.__chat.acreate).acreate(*args, **kwargs)
+
+
+# This wraps 0.*.* versions of the openai module, eg https://github.com/openai/openai-python/tree/v0.28.1
+class OpenAIV0Wrapper(NamedWrapper):
+    def __init__(self, openai):
+        super().__init__(openai)
+        self.ChatCompletion = ChatCompletionV0Wrapper(openai.ChatCompletion)
+
+
+class CompletionsV1Wrapper(NamedWrapper):
+    def __init__(self, completions):
+        self.__completions = completions
+        super().__init__(completions)
+
+    def create(self, *args, **kwargs):
+        return ChatCompletionWrapper(self.__completions.create, None).create(*args, **kwargs)
+
+
+class AsyncCompletionsV1Wrapper(NamedWrapper):
+    def __init__(self, completions):
+        self.__completions = completions
+        super().__init__(completions)
+
+    async def create(self, *args, **kwargs):
+        return await ChatCompletionWrapper(None, self.__completions.create).acreate(*args, **kwargs)
+
+
+class ChatV1Wrapper(NamedWrapper):
+    def __init__(self, chat):
+        super().__init__(chat)
 
-def run_cached_request(Completion=None, **kwargs):
-    if Completion is None:
-        # OpenAI is very slow to import, so we only do it if we need it
         import openai
```

Comment:
we don't need this import anymore?

Reply:
we do on the RHS because we do a typecheck